### PR TITLE
[AIRFLOW-3327] Add support for location in BigQueryHook

### PR DESCRIPTION
**Pull request by kaxil**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3327

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

The geographic location of the job is required except for US and EU. See details at https://cloud.google.com/bigquery/docs/locations#specifying_your_location.

This PR also cleans some tests

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Comment by kaxil** [comment]
cc @fenglu-g

File: airflow/contrib/hooks/bigquery_hook.py [hunk 1]
```diff
@@ -580,11 +587,18 @@ def run_query(self,
             by one or more columns. This is only available in combination with
             time_partitioning. The order of columns given determines the sort order.
         :type cluster_fields: list of str
+        :param location: The geographic location of the job. Required except for
+            US and EU. See details at
```
Comment (XD-DENG) [inline airflow/contrib/hooks/bigquery_hook.py:591]: Hi @kaxil , curious what would happen if the BigQuery data is stored in a region location other than US/EU while user misses to specify `location`? How will the exception be like?


`

Comment (kaxil) [reply airflow/contrib/hooks/bigquery_hook.py:591]: They would get the following error if the data is in for example `asia-northeast1` (Japan) location and they don't specify the location:

```
Error executing query job. Message: 404 Not Found
{
  "code" : 404,
  "errors" : [ {
    "domain" : "global",
    "message" : "Not found: Dataset mybqproject:mybqtable",
    "reason" : "notFound"
  } ],
  "message" : "Not found: Dataset mybqproject:mybqtable"
},
```

**Comment by kaxil** [comment]
cc @dsdinter @tswast

File: airflow/contrib/hooks/bigquery_hook.py (hunk 1 above)
Comment (yohei1126) [reply airflow/contrib/hooks/bigquery_hook.py:591]: Hi, I tried to import to a data set in `asia-northeast1` using `GoogleCloudStorageToBigQueryOperator ` and I got the following error as kaxil mentioned.
```[2018-12-26 21:35:47,464] {base_task_runner.py:107} INFO - Job 146: Subtask bq_load_data_into_dest_table_from_gcs [2018-12-26 21:35:47,464] {discovery.py:871} INFO - URL being requested: GET https://www.googleapis.com/bigquery/v2/projects/my-project/jobs/job_abc123?alt=json
[2018-12-26 21:35:47,931] {models.py:1736} ERROR - ('BigQuery job status check failed. Final error was: %s', 404)
Traceback (most recent call last)
  File "/usr/local/lib/airflow/airflow/contrib/hooks/bigquery_hook.py", line 981, in run_with_configuratio
    jobId=self.running_job_id).execute(
  File "/usr/local/lib/python3.6/site-packages/googleapiclient/_helpers.py", line 130, in positional_wrappe
    return wrapped(*args, **kwargs
  File "/usr/local/lib/python3.6/site-packages/googleapiclient/http.py", line 851, in execut
    raise HttpError(resp, content, uri=self.uri
googleapiclient.errors.HttpError: <HttpError 404 when requesting https://www.googleapis.com/bigquery/v2/projects/my-project/jobs/job_abc123?alt=json returned "Not found: Job my-project:job_abc123"

During handling of the above exception, another exception occurred

Traceback (most recent call last)
  File "/usr/local/lib/airflow/airflow/models.py", line 1633, in _run_raw_tas
    result = task_copy.execute(context=context
  File "/usr/local/lib/airflow/airflow/contrib/operators/gcs_to_bq.py", line 237, in execut
    time_partitioning=self.time_partitioning
  File "/usr/local/lib/airflow/airflow/contrib/hooks/bigquery_hook.py", line 951, in run_loa
    return self.run_with_configuration(configuration
  File "/usr/local/lib/airflow/airflow/contrib/hooks/bigquery_hook.py", line 1003, in run_with_configuratio
    err.resp.status
Exception: ('BigQuery job status check failed. Final error was: %s', 404
```
https://issues.apache.org/jira/browse/AIRFLOW-3571

Comment (kaxil) [reply airflow/contrib/hooks/bigquery_hook.py:591]: @yohei1126 Thanks for confirming, can you also try out with the changes in the PR and see if it fixes the issue for you? 

Once you confirm, I will then merge it.

Comment (yohei1126) [reply airflow/contrib/hooks/bigquery_hook.py:591]: OK will do tomorrow

Comment (yohei1126) [reply airflow/contrib/hooks/bigquery_hook.py:591]: This operator has `destination_dataset_table` so I tested the following four patterns. I suppose it works as expected.

1. (OK) source dataset in US, destination_dataset in US, no location specified
2. (OK) source dataset in Tokyo, destination_dataset in Tokyo, location specified as `asia-northeast1`
3. (Fail) source dataset in Tokyo, destination_dataset in US, location specified as `asia-northeast1`
4. (Fail) source dataset in US, destination_dataset in Tokyo, no location specified


```
# -*- coding: utf-8 -*-
from airflow import DAG
from operators.bigquery_operator import BigQueryOperator
from airflow.utils.dates import days_ago

ARGS = {
    'owner': 'airflow',
    'depends_on_past': False,
    'retries': 0,
    'start_date': days_ago(1)
}

COMMON_PARAMS = {}

with DAG(
    dag_id='test_bq',
    default_args=ARGS,
    params=COMMON_PARAMS,
    schedule_interval='@once') as dag:

    # both datasets are in US
    # this should be OK
    t1 = BigQueryOperator(
        task_id='us_to_us',
        sql='SELECT * FROM test_us_ds.test_table',
        bigquery_conn_id='google_cloud_default',
        destination_dataset_table='fr-stg-datalake:test_us_ds.dest_table_us_to_us'
    )

    # both source dataset and dest dataset are in Tokyo
    # this should be OK
    t2 = BigQueryOperator(
        task_id='tky_to_tky',
        sql='SELECT * FROM test_tokyo_ds.test_table',
        bigquery_conn_id='google_cloud_default',
        location='asia-northeast1',
        destination_dataset_table='fr-stg-datalake:test_tokyo_ds.dest_table_tky_to_tky'
    )

    # source dataset is in Tokyo but dest table is in US
    # this should fail
    t3 = BigQueryOperator(
        task_id='tky_to_us',
        sql='SELECT * FROM test_tokyo_ds.test_table',
        bigquery_conn_id='google_cloud_default',
        location='asia-northeast1',
        destination_dataset_table='fr-stg-datalake:test_us_ds.dest_table_tky_to_us'
    )

    # source dataset is in US but dest table is in Tokyo
    # this should fail
    t4 = BigQueryOperator(
        task_id='us_to_tky',
        sql='SELECT * FROM test_us_ds.test_table',
        bigquery_conn_id='google_cloud_default',
        destination_dataset_table='fr-stg-datalake:test_tokyo_ds.dest_table_us_to_tky'
    )

    t1 >> t2
    t1 >> t3
    t1 >> t4
    globals()[dag] = dag
```

Comment (yohei1126) [reply airflow/contrib/hooks/bigquery_hook.py:591]: FYI: I downloaded `bigquery_operator.py` and `bigquery_hook.py` from this PR and uploaded to my composer (location = `asia-northeast1` ).

Comment (kaxil) [reply airflow/contrib/hooks/bigquery_hook.py:591]: Great, thanks for testing it. I will merge this today.

**Review by tswast — APPROVED** [review]

File: tests/contrib/hooks/test_bigquery_hook.py [hunk 2]
```diff
@@ -729,5 +702,21 @@ def test_legacy_sql_override_propagates_properly(self, run_with_config):
...
+    def test_location_propagates_properly(self, run_with_config):
+        with mock.patch.object(hook.BigQueryHook, 'get_service'):
+            bq_hook = hook.BigQueryHook(location=None)
+            self.assertIsNone(bq_hook.location)
+
+            bq_cursor = hook.BigQueryBaseCursor(mock.Mock(),
+                                                'test-project',
+                                                location=None)
+            self.assertIsNone(bq_cursor.location)
+            bq_cursor.run_query(sql='select 1', location='US')
```
Comment (yohei1126) [inline tests/contrib/hooks/test_bigquery_hook.py]: Why do you need to specify `location` on `run_query` after initialization of `BigQueryHook`?

Comment (kaxil) [reply tests/contrib/hooks/test_bigquery_hook.py]: It gives an option to override defaults

**Pull request merged**